src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,451 @@
+update_machine_property(JsonObject *object,
+                        const gchar *member_name,
+                        JsonNode *member_node,
+                        gpointer user_data)
+{
+  json_object_set_member ((JsonObject*) user_data, member_name, json_node_copy (member_node));
+}
+
+static gboolean
+update_machine (const char *filename, const char* hostname, JsonNode *info, GError **error)
+{
+  gchar *path;
+  JsonNode *cur_config;
+  JsonGenerator *json_gen;
+  JsonObject *cur_config_obj;
+  JsonNode *cur_props;
+  gboolean res;
+
+  g_assert (JSON_NODE_HOLDS_OBJECT (info));
+
+  path = g_build_filename (machines_json_dir(), filename, NULL);
+  cur_config = parse_json_file (path);
+  if (cur_config == NULL)
+    cur_config = new_object_node ();
+  cur_config_obj = json_node_get_object (cur_config);
+  cur_props = json_object_get_member (cur_config_obj, hostname);
+
+  if (cur_props)
+    {
+      /* update settings for hostname */
+      g_assert (JSON_NODE_HOLDS_OBJECT(cur_props));
+      json_object_foreach_member (json_node_get_object (info), update_machine_property, json_node_get_object (cur_props));
+    }
+  else
+    /* create new entry for host name */
+    json_object_set_member (cur_config_obj, hostname, json_node_copy (info));
+
+  /* write back json file */
+  json_gen = json_generator_new ();
+  json_generator_set_root (json_gen, cur_config);
+  json_generator_set_pretty (json_gen, TRUE); /* bikeshed zone */
+  res = json_generator_to_file (json_gen, path, error);
+  g_free (path);
+  g_object_unref (json_gen);
+  json_node_free (cur_config);
+  return res;
+}
+
+static GVariant *
+machines_get_property (GDBusConnection *connection,
+                       const gchar *sender,
+                       const gchar *object_path,
+                       const gchar *interface_name,
+                       const gchar *property_name,
+                       GError **error,
+                       gpointer user_data)
+{
+  g_return_val_if_fail (property_name != NULL, NULL);
+
+  if (g_str_equal (property_name, "Machines"))
+    return get_machines ();
+  else
+    g_return_val_if_reached (NULL);
+}
+
+static void
+machines_method_call (GDBusConnection *connection,
+                      const gchar *sender,
+                      const gchar *object_path,
+                      const gchar *interface_name,
+                      const gchar *method_name,
+                      GVariant *parameters,
+                      GDBusMethodInvocation *invocation,
+                      gpointer user_data)
+{
+  if (g_str_equal (method_name, "Update"))
+    {
+      const gchar *filename, *hostname;
+      GVariant * info_v;
+      JsonNode *info_json = NULL;
+      GError *error = NULL;
+
+      g_variant_get (parameters, "(&s&s@a{sv})", &filename, &hostname, &info_v);
+      info_json = json_gvariant_serialize (info_v);
+      /* g_debug ("Updating %s for machine %s: %s", filename, hostname, g_variant_print (info_v, TRUE)); */
```

Comment:
I left this commented as it causes a memleak (`g_variant_print()` result needs to be freed and thus also needs a new variable) and it's relatively expensive. I'll uncomment it but leave out the value of the gvariant, it was mostly just necessary for developing the correct disassembling of the variant.